common/src/main/java/io/netty/util/ResourceLeakDetector.java

```diff
@@ -453,54 +503,52 @@ public String toString() {
         }
     }
 
-    private static final String[] STACK_TRACE_ELEMENT_EXCLUSIONS = {
-            "io.netty.util.ReferenceCountUtil.touch(",
-            "io.netty.buffer.AdvancedLeakAwareByteBuf.touch(",
-            "io.netty.buffer.AbstractByteBufAllocator.toLeakAwareBuffer(",
-            "io.netty.buffer.AdvancedLeakAwareByteBuf.recordLeakNonRefCountingOperation("
-    };
-
-    static String newRecord(Object hint, int recordsToSkip) {
-        StringBuilder buf = new StringBuilder(4096);
-
-        // Append the hint first if available.
-        if (hint != null) {
-            buf.append("\tHint: ");
-            // Prefer a hint string to a simple string form.
-            if (hint instanceof ResourceLeakHint) {
-                buf.append(((ResourceLeakHint) hint).toHintString());
-            } else {
-                buf.append(hint);
-            }
-            buf.append(NEWLINE);
+    private static final class Record extends Throwable {
+
+        private static final Set<String> STACK_TRACE_ELEMENT_EXCLUSIONS = new HashSet<String>();
+        static {
+            STACK_TRACE_ELEMENT_EXCLUSIONS.add("io.netty.util.ReferenceCountUtil.touch");
+            STACK_TRACE_ELEMENT_EXCLUSIONS.add("io.netty.buffer.AdvancedLeakAwareByteBuf.touch");
+            STACK_TRACE_ELEMENT_EXCLUSIONS.add("io.netty.buffer.AbstractByteBufAllocator.toLeakAwareBuffer");
+            STACK_TRACE_ELEMENT_EXCLUSIONS.add(
+                    "io.netty.buffer.AdvancedLeakAwareByteBuf.recordLeakNonRefCountingOperation");
         }
 
-        // Append the stack trace.
-        StackTraceElement[] array = new Throwable().getStackTrace();
-        for (StackTraceElement e: array) {
-            if (recordsToSkip > 0) {
-                recordsToSkip --;
-            } else {
-                String estr = e.toString();
+        private final String hintString;
+        Record next;
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Make this "next" field final. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1165)